src/gc/gc.cpp

```diff
@@ -12327,33 +12332,44 @@ BOOL gc_heap::a_fit_segment_end_p (int gen_number,
     }
 #endif //BACKGROUND_GC
 
-    uint8_t* old_alloc;
-    old_alloc = allocated;
 #ifdef FEATURE_LOH_COMPACTION
     if (gen_number == (max_generation + 1))
     {
-        make_unused_array (old_alloc, loh_pad);
-        old_alloc += loh_pad;
+        make_unused_array (allocated, loh_pad);
         allocated += loh_pad;
         limit -= loh_pad;
     }
 #endif //FEATURE_LOH_COMPACTION
 
-#if defined (VERIFY_HEAP) && defined (_DEBUG)
-        ((void**) allocated)[-1] = 0;     //clear the sync block
-#endif //VERIFY_HEAP && _DEBUG
-    allocated += limit;
-
     dprintf (3, ("found fit at end of seg: %Ix", old_alloc));
 
+    uint8_t* old_alloc;
+    old_alloc = allocated;
+
 #ifdef BACKGROUND_GC
     if (cookie != -1)
     {
+        allocated += limit;
         bgc_loh_alloc_clr (old_alloc, limit, acontext, flags, align_const, cookie, TRUE, seg);
     }
     else
 #endif //BACKGROUND_GC
-    {
+    {      
+        // In a contiguous AC case with GC_ALLOC_ZEROING_OPTIONAL, deduct unspent space from the limit to clear only what is necessary.
+        if (flags & GC_ALLOC_ZEROING_OPTIONAL &&
+            (allocated == acontext->alloc_limit || allocated == acontext->alloc_limit + Align (min_obj_size, align_const)))
```

Comment:
>allocated == acontext->alloc_limit [](start = 13, length = 34)

nit - as a convention in GC code, please always put ()'s around individual statements -

        if ((flags & GC_ALLOC_ZEROING_OPTIONAL) &&
            ((allocated == acontext->alloc_limit) || (allocated == (acontext->alloc_limit + Align (min_obj_size, align_const)))))

Reply:
I am trying :-) , I think I missed couple ( ) here since there is a lot of nesting. 

---
In reply to: [287172722](https://github.com/dotnet/coreclr/pull/24504#discussion_r287172722) [](ancestors = 287172722)